File: BlueSTSDK/BlueSTSDK/Util/BlueSTSDKBleNodeDefines.m
/*******************************************************************************
 * COPYRIGHT(c) 2015 STMicroelectronics
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *   1. Redistributions of source code must retain the above copyright notice,
 *      this list of conditions and the following disclaimer.
 *   2. Redistributions in binary form must reproduce the above copyright notice,
 *      this list of conditions and the following disclaimer in the documentation
 *      and/or other materials provided with the distribution.
 *   3. Neither the name of STMicroelectronics nor the names of its contributors
 *      may be used to endorse or promote products derived from this software
 *      without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 ******************************************************************************/

#import <Foundation/Foundation.h>

#import "../Features/BlueSTSDKFeatureAcceleration.h"
#import "../Features/BlueSTSDKFeatureGyroscope.h"
#import "../Features/BlueSTSDKFeatureMagnetometer.h"
#import "../Features/BlueSTSDKFeatureMemsSensorFusion.h"
#import "../Features/BlueSTSDKFeatureHumidity.h"
#import "../Features/BlueSTSDKFeatureTemperature.h"
#import "../Features/BlueSTSDKFeaturePressure.h"
#import "../Features/BlueSTSDKFeatureMemsSensorFusionCompact.h"
#import "../Features/BlueSTSDKFeatureLuminosity.h"
#import "../Features/BlueSTSDKFeatureProximity.h"
#import "../Features/BlueSTSDKFeatureBattery.h"
#import "../Features/BlueSTSDKFeatureFreeFall.h"
#import "../Features/BlueSTSDKFeatureCarryPosition.h"
#import "../Features/BlueSTSDKFeatureMicLevel.h"
#import "../Features/BlueSTSDKFeatureProximityGesture.h"
#import "../Features/BlueSTSDKFeatureMemsGesture.h"
#import "../Features/BlueSTSDKFeaturePedometer.h"
#import "../Features/BlueSTSDKFeatureAccelerometerEvent.h"
#import "../Features/BlueSTSDKFeatureDirectionOfArrival.h"
#import "../Features/BlueSTSDKFeatureSwitch.h"
#import "../Features/BlueSTSDKFeatureCompass.h"
#import "../Features/Remote/BlueSTSDKRemoteFeatureSwitch.h"
#import "../Features/Remote/BlueSTSDKRemoteFeatureTemperature.h"
#import "../Features/Remote/BlueSTSDKRemoteFeaturePressure.h"
#import "../Features/Remote/BlueSTSDKRemoteFeatureHumidity.h"
#import "../Features/BlueSTSDKFeatureAudioADPCM.h"
#import "../Features/BlueSTSDKFeatureAudioADPCMSync.h"
#import "../Features/BlueSTSDKFeatureMotionIntensity.h"
#import "../Features/BlueSTSDKFeatureSDLogging.h"
#import "../Features/BlueSTSDKFeatureCOSensor.h"
#import "BlueSTSDK/BlueSTSDK-Swift.h"

#import "BlueSTSDKBleNodeDefines.h"
#import "NSData+NumberConversion.h"
#import "BlueSTSDKFeatureBeamForming.h"

//all the sdk characteristics will end with this string
#define COMMON_CHAR_UUID @"-11E1-AC36-0002A5D5C51B"
//all the features characteristics exported in the advertise feature mask will end with this string
#define BASE_FEATURE_COMMON_UUID @"-0001" COMMON_CHAR_UUID
//all the features characteristics not exported will end with this string
#define EXTENDED_FEATURE_COMMON_UUID @"-0002" COMMON_CHAR_UUID
//all the general purpose characteristics will end with this string
#define COMMON_GP_FEATURE_UUID @"-0003" COMMON_CHAR_UUID
//all the sdk service will end with this string
#define COMMON_SERVICE_UUID "-11E1-9AB4-0002A5D5C51B"

static NSDictionary<CBUUID*,NSArray<Class>*> *EXTENDED_FEATURE_MAP = nil;


@implementation BlueSTSDKFeatureCharacteristics

+(CBUUID*) buildExtendedFeatureCharacteristicsWithPrefix:(uint32_t) prefix{
    NSString *uuidString = [NSString stringWithFormat:@"%08X%@",prefix,EXTENDED_FEATURE_COMMON_UUID];
    return [CBUUID UUIDWithString: uuidString];
}
/*
 EXTENDED_FEATURE_MAP.put(buildExtendedFeatureCharacteristics(0x07), asList(FeaturePredictiveSpeedStatus.class));
 EXTENDED_FEATURE_MAP.put(buildExtendedFeatureCharacteristics(0x08), asList(FeaturePredictiveAccelerationStatus.class));
 EXTENDED_FEATURE_MAP.put(buildExtendedFeatureCharacteristics(0x09), asList(FeaturePredictiveFrequencyDomainStatus.class));
 */
+(void)initialize{
    if(self == [BlueSTSDKFeatureCharacteristics class]){
        EXTENDED_FEATURE_MAP = @{
                                 [BlueSTSDKFeatureCharacteristics buildExtendedFeatureCharacteristicsWithPrefix: 0x3]:@[[BlueSTSDKFeatureAudioSceneCalssification class]],
                                 [BlueSTSDKFeatureCharacteristics buildExtendedFeatureCharacteristicsWithPrefix: 0x4]:@[[BlueSTSDKFeatureAILogging class]],
                                 [BlueSTSDKFeatureCharacteristics buildExtendedFeatureCharacteristicsWithPrefix: 0x5]:@[[BlueSTSDKFeatureFFTAmplitude class]],
                                 [BlueSTSDKFeatureCharacteristics buildExtendedFeatureCharacteristicsWithPrefix: 0x6]:@[[BlueSTSDKFeatureMotorTimeParameters class]],
                                 [BlueSTSDKFeatureCharacteristics buildExtendedFeatureCharacteristicsWithPrefix: 0x7]:@[[BlueSTSDKFeaturePredictiveSpeedStatus class]],
                                 [BlueSTSDKFeatureCharacteristics buildExtendedFeatureCharacteristicsWithPrefix: 0x8]:@[[BlueSTSDKFeaturePredictiveAccelerationStatus class]],
                                 [BlueSTSDKFeatureCharacteristics buildExtendedFeatureCharacteristicsWithPrefix: 0x9]:@[[BlueSTSDKFeaturePredictiveFrequencyDomainStatus class]],
                                 };
    }
    
}

+(featureMask_t)extractFeatureMask:(CBUUID *)uuid{
    return [uuid.data extractBeUInt32FromOffset:0];
}//extractFeatureMask

+(bool) isFeatureCharacteristics:(CBCharacteristic*) c{
    return [ c.UUID.UUIDString hasSuffix:BASE_FEATURE_COMMON_UUID];
}//isFeatureCharacteristics

+(bool) isFeatureGeneralPurposeCharacteristics:(CBCharacteristic*) c{
    return [ c.UUID.UUIDString hasSuffix:COMMON_GP_FEATURE_UUID];
}//isFeatureGeneralPurposeCharacteristics

+(nullable NSArray<Class>*)getExtendedFeatureInCharacteristics:(CBCharacteristic*) c{
    return EXTENDED_FEATURE_MAP[c.UUID];
}

@end

#define CONFIG_SERVICE_ID "000F"
#define CONFIG_SERVICE_UUID @"00000000-"CONFIG_SERVICE_ID COMMON_SERVICE_UUID
#define CONFIG_CONTROL_CHAR_UUID @"00000001-"CONFIG_SERVICE_ID COMMON_CHAR_UUID
#define COMMAND_CONFIG_CHAR_UUID @"00000002-"CONFIG_SERVICE_ID COMMON_CHAR_UUID

@implementation BlueSTSDKServiceConfig
+(CBUUID*) serviceUuid{
    static CBUUID *service = nil;
    if(service==nil)
        service=[CBUUID UUIDWithString:CONFIG_SERVICE_UUID ];
    return service;
}//serviceUuid

+(CBUUID*) configControlUuid{
    static CBUUID *configControl =nil;
    if(configControl ==nil)
        configControl = [CBUUID UUIDWithString: CONFIG_CONTROL_CHAR_UUID];
    return configControl;
}//configControlUuid

+(CBUUID*) featureCommandUuid{
    static CBUUID *commandFeature =nil;
    if(commandFeature ==nil)
        commandFeature = [CBUUID UUIDWithString: COMMAND_CONFIG_CHAR_UUID];
    return commandFeature;
}//featureCommandUuid

+(bool) isConfigCharacteristics:(CBCharacteristic*) c {
    return [c.UUID isEqual: BlueSTSDKServiceConfig.configControlUuid] ||
    [c.UUID isEqual: BlueSTSDKServiceConfig.featureCommandUuid];
}//isConfigCharacteristics

+(bool) isConfigControlCharacteristic:(CBCharacteristic*) c {
    return [c.UUID isEqual: BlueSTSDKServiceConfig.configControlUuid];
}//isConfigControlCharacteristic

+(bool) isConfigFeatureCommandCharacteristic:(CBCharacteristic*) c {
    return [c.UUID isEqual: BlueSTSDKServiceConfig.featureCommandUuid];
}//isConfigFeatureCommandCharacteristic

@end

#define DEBUG_SERVICE_ID "000E"
#define DEBUG_SERVICE_UUID @"00000000-"DEBUG_SERVICE_ID COMMON_SERVICE_UUID
#define TEMR_DEBUG_CHAR_UUID @"00000001-"DEBUG_SERVICE_ID COMMON_CHAR_UUID
#define STDERR_DEBUG_CHAR_UUID @"00000002-"DEBUG_SERVICE_ID COMMON_CHAR_UUID

@implementation BlueSTSDKServiceDebug
+(CBUUID*) serviceUuid{
    static CBUUID *service = nil;
    if(service==nil)
        service=[CBUUID UUIDWithString:DEBUG_SERVICE_UUID ];
    return service;
}//serviceUuid

+(CBUUID*) stdErrUuid{
    static CBUUID *stdErr =nil;
    if(stdErr ==nil)
        stdErr = [CBUUID UUIDWithString: STDERR_DEBUG_CHAR_UUID];
    return stdErr;
}//stdErrUuid

+(CBUUID*) termUuid{
    static CBUUID *term =nil;
    if(term ==nil)
        term = [CBUUID UUIDWithString: TEMR_DEBUG_CHAR_UUID];
    return term;
}//termUuid

+(bool) isDebugCharacteristics:(CBCharacteristic*) c{
    return [c.UUID isEqual: BlueSTSDKServiceDebug.termUuid] ||
    [c.UUID isEqual: BlueSTSDKServiceDebug.stdErrUuid];
}//isDebugCharacteristics

@end

@implementation BlueSTSDKServices

+(BlueSTSDKServiceDebug*) debugService{
    static BlueSTSDKServiceDebug *service = nil;
    if(service==nil)
        service=[BlueSTSDKServiceDebug alloc];
    return service;
}//debugService

+(BlueSTSDKServiceConfig*) configService{
    static BlueSTSDKServiceConfig *service = nil;
    if(service==nil)
        service=[BlueSTSDKServiceConfig alloc];
    return service;
}//configService

@end

File: BlueSTSDK/BlueSTSDK/Util/BlueSTSDKBleNodeDefines.swift
/*******************************************************************************
 * COPYRIGHT(c) 2019 STMicroelectronics
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *   1. Redistributions of source code must retain the above copyright notice,
 *      this list of conditions and the following disclaimer.
 *   2. Redistributions in binary form must reproduce the above copyright notice,
 *      this list of conditions and the following disclaimer in the documentation
 *      and/or other materials provided with the distribution.
 *   3. Neither the name of STMicroelectronics nor the names of its contributors
 *      may be used to endorse or promote products derived from this software
 *      without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 ******************************************************************************/

import Foundation
import CoreBluetooth

public typealias featureMask_t = UInt32

@objc public //todo add internal visibity when remove the objc annotation
extension CBUUID {
    /**
     *  extract the feature mask from an uuid value, tell to the user witch feature
     *  are exported by this characteristics
     *
     *  @return characteristics feature mask
     */
    var featureMask:featureMask_t{
        get{
            return (self.data as NSData).extractBeUInt32(fromOffset: 0)
        }
    }
}

@objc public //todo add internal visibity when remove the objc annotation
extension CBCharacteristic {
    
    //all the sdk characteristics will end with this string
    fileprivate static let COMMON_CHAR_UUID = "-11E1-AC36-0002A5D5C51B"
    //all the features characteristics exported in the advertise feature mask will end with this string
    private static  let BASE_FEATURE_COMMON_UUID = "-0001"+COMMON_CHAR_UUID
    //all the features characteristics not exported will end with this string
    private static  let  EXTENDED_FEATURE_COMMON_UUID = "-0002"+COMMON_CHAR_UUID
    //all the general purpose characteristics will end with this string
    private static  let COMMON_GP_FEATURE_UUID = "-0003"+COMMON_CHAR_UUID
    //all the sdk service will end with this string
    private static  let COMMON_SERVICE_UUID = "-11E1-9AB4-0002A5D5C51B"
    
    /**
     *  tell if a characteristics has a valid uuid for be manage by this sdk
     *
     *  @param c characteristic to test
     *
     *  @return true if the characteristics can be manage by this sdk
     */
    var isFeatureCaracteristics:Bool{
        get {
            return self.uuid.uuidString.hasSuffix(CBCharacteristic.BASE_FEATURE_COMMON_UUID)
        }
    }
    
    /**
     *  tell if a characteristics has a valid uudi for be manage by this sdk as a
     *  general pourpose characteristics
     *
     *  @param c characteristic to test
     *
     *  @return true if the characteristics can be manage by this sdk as general purpose characteristics
     */
    var isFeatureGeneralPurposeCharacteristics:Bool{
        get{
            return self.uuid.uuidString.hasSuffix(CBCharacteristic.COMMON_GP_FEATURE_UUID)
        }
    }
    
    private static func buildExtendedFeatureCharacteristics( prefix:UInt32 )->CBUUID{
        let uuidString = String(format: "%08X%@", prefix,CBCharacteristic.EXTENDED_FEATURE_COMMON_UUID)
        return CBUUID(string: uuidString)
    }
    
    private static let EXTENDED_FEATURE_MAP = [
        buildExtendedFeatureCharacteristics(prefix: 0x01) : [BlueSTSDKFeatureAudioOpus.self],
        buildExtendedFeatureCharacteristics(prefix: 0x02) : [BlueSTSDKFeatureAudioOpusConf.self],
        buildExtendedFeatureCharacteristics(prefix: 0x03) : [BlueSTSDKFeatureAudioCalssification.self],
        buildExtendedFeatureCharacteristics(prefix: 0x04) : [BlueSTSDKFeatureAILogging.self],
        buildExtendedFeatureCharacteristics(prefix: 0x05) : [BlueSTSDKFeatureFFTAmplitude.self],
        buildExtendedFeatureCharacteristics(prefix: 0x06) : [BlueSTSDKFeatureMotorTimeParameters.self],
        buildExtendedFeatureCharacteristics(prefix: 0x07) : [BlueSTSDKFeaturePredictiveSpeedStatus.self],
        buildExtendedFeatureCharacteristics(prefix: 0x08) : [BlueSTSDKFeaturePredictiveAccelerationStatus.self],
        buildExtendedFeatureCharacteristics(prefix: 0x09) : [BlueSTSDKFeaturePredictiveFrequencyDomainStatus.self],
        buildExtendedFeatureCharacteristics(prefix: 0x0A) : [BlueSTSDKFeatureMotionAlogrithm.self],
        //buildExtendedFeatureCharacteristics(prefix: 0x0B) : [],
        //buildExtendedFeatureCharacteristics(prefix: 0x0C) : [],
        buildExtendedFeatureCharacteristics(prefix: 0x0D) : [BlueSTSDKFeatureEulerAngle.self],
        buildExtendedFeatureCharacteristics(prefix: 0x0E) : [BlueSTSDKFeatureFitnessActivity.self],
    ]
    
    var extendedFeature:[BlueSTSDKFeature.Type]?{
        get{
            return CBCharacteristic.EXTENDED_FEATURE_MAP[self.uuid]
        }
    }
}

fileprivate //todo add internal visibity when remove the objc annotation, and made it a struct
struct BlueSTSDKService {
    //all the sdk service will end with this string
    private static let COMMON_SERVICE_UUID = "-11E1-9AB4-0002A5D5C51B"
    
    /**
     *  class with the utility function for the debug service
     */
    struct BlueSTSDKDebugService {
        private static let SERVICE_ID  = "000E"
        let uuid = CBUUID(string: "00000000-\(BlueSTSDKDebugService.SERVICE_ID)\(BlueSTSDKService.COMMON_SERVICE_UUID)")
        let termUuid = CBUUID(string: "00000001-\(BlueSTSDKDebugService.SERVICE_ID)\(CBCharacteristic.COMMON_CHAR_UUID)")
        let stdErrUuid = CBUUID(string: "00000002-\(BlueSTSDKDebugService.SERVICE_ID)\(CBCharacteristic.COMMON_CHAR_UUID)")
    }
    
    struct BlueSTSDKConfigService {
        private static let SERVICE_ID  = "000F"
        let uuid = CBUUID(string: "00000000-\(BlueSTSDKConfigService.SERVICE_ID)\(BlueSTSDKService.COMMON_SERVICE_UUID)")
        let configControlUuid = CBUUID(string: "00000001-\(BlueSTSDKConfigService.SERVICE_ID)\(CBCharacteristic.COMMON_CHAR_UUID)")
        let featureCommandUuid = CBUUID(string: "00000002-\(BlueSTSDKConfigService.SERVICE_ID)\(CBCharacteristic.COMMON_CHAR_UUID)")

        
    }
    
    static let debugService = BlueSTSDKDebugService()
    static let configService = BlueSTSDKConfigService()
}

@objc public //todo add internal visibity when remove the objc annotation
extension CBService {
    
    var isDebugService:Bool{
        get{
            return self.uuid == BlueSTSDKService.debugService.uuid
        }
    }
    
    var isConfigService:Bool{
        get{
            return self.uuid == BlueSTSDKService.configService.uuid
        }
    }
}

@objc public //todo add internal visibity when remove the objc annotation
extension CBCharacteristic {
    var isDebugCharacteristic:Bool{
        get{
            return isDebugTermCharacteristic || isDebugErrorCharacteristic
        }
    }
    var isDebugTermCharacteristic:Bool{
        get{
            return self.uuid == BlueSTSDKService.debugService.termUuid
        }
    }
    
    var isDebugErrorCharacteristic:Bool{
        get{
            return self.uuid == BlueSTSDKService.debugService.stdErrUuid
        }
    }
}

@objc public //todo add internal visibity when remove the objc annotation
extension CBCharacteristic {
    var isConfigCharacteristics:Bool{
        get{
            return isConfigControlCharacteristic || isConfigFeatureCommandCharacteristic
        }
    }
    
    var isConfigControlCharacteristic:Bool{
        get{
            return self.uuid == BlueSTSDKService.configService.configControlUuid
        }
    }
    
    var isConfigFeatureCommandCharacteristic:Bool{
        get{
            return self.uuid == BlueSTSDKService.configService.featureCommandUuid
        }
    }
}
